libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py

```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
+                    f"Error occurs by inference endpoint: {str(err)}"
+                ) from err
+
+        return _completion_with_retry(**kwargs)
+
+    async def acompletion_with_retry(
+        self,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> Any:
+        """Use tenacity to retry the async completion call."""
+        retry_decorator = _create_retry_decorator(self, run_manager=run_manager)
+
+        @retry_decorator
+        async def _completion_with_retry(**kwargs: Any) -> Any:
+            try:
+                request_timeout = kwargs.pop("request_timeout", DEFAULT_TIME_OUT)
+                data = kwargs.pop("data")
+                stream = kwargs.pop("stream", self.streaming)
+
+                request = Requests(headers=self._headers(is_async=True, body=data))
+                if stream:
+                    response = request.apost(
+                        url=self.endpoint,
+                        data=data,
+                        timeout=request_timeout,
+                    )
+                    return self._aiter_sse(response)
+                else:
+                    async with request.apost(
+                        url=self.endpoint,
+                        data=data,
+                        timeout=request_timeout,
+                    ) as response:
+                        self._check_response(response)
+                        data = await response.json()
+                        return data
+            except TokenExpiredError as e:
+                raise e
+            except Exception as err:
+                logger.debug(
+                    f"Requests payload: `{data}`. "
+                    f"Stream mode={stream}. "
+                    f"Requests kwargs: url={self.endpoint}, timeout={request_timeout}."
+                )
+                raise ValueError(
+                    f"Error occurs by inference endpoint: {str(err)}"
+                ) from err
+
+        return await _completion_with_retry(**kwargs)
+
+    def _check_response(
+        self, response: Union[requests.Response, aiohttp.ClientResponse]
+    ) -> None:
+        """Handle server error by checking the response status.
+
+        Args:
+            response (Union[requests.Response, aiohttp.ClientResponse]):
+                The response object from either `requests` or `aiohttp` library.
+
+        Raises:
+            TokenExpiredError:
+                If the response status code is 401 and the token refresh is successful.
+            ServerError:
+                If any other HTTP error occurs.
+        """
+        try:
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as http_err:
+            status_code = (
+                response.status_code
+                if hasattr(response, "status_code")
+                else response.status
+            )
+            if status_code == 401 and self._refresh_signer():
```

Comment:
Can you elaborate on this logic please?

Reply:
sure, this is refactoring this part: https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py#L181-L191 . Allowing retry for refresh token before invoking.